text/0000-introduce empty ranges.md

```diff
@@ -0,0 +1,431 @@
+- Feature Name: Introduction of empty Ranges
+- Start Date: 2022-05-29
+- RFC PR: (leave this empty)
+- Pony Issue: (leave this empty)
+
+# Summary
+
+The RFC addresses the fact that for certain parameters, Pony currently creates infinite Range iterators that produce infinite numbers of values *outside* the range of values `min` and `max` provided to `Range`. For example the Range iterator `Range(10, 0)` produces the infinite sequence `10, 11, 12, ..` on subsequent calls to `.next()` where all but the first element do not lie within the range 10..0. One unexpected side effect of Pony's current behavior is that even though finite bounds are given, indefinitely-running actors and "hanging" programs can be produced. The fact that this RFC would bring Pony's Range much more in line with other common implementations of Range iterators that do allow floating point parameters and even infinite arguments -- for example Numpy's `arange()` function -- can be seen as a validating circumstance, but a more important goal is to prevent unintended infinitely iterating Range objects and indefinitely-running actors.
+
+It is argued here that an iterator using the name `Range` should produce a sequence of values that lie within the `min, max` bounds. Therefore, most of what are now *infinite* ranges should instead be *empty* ranges (where `.has_next() == false`). The RFC proposes a delineation of the so-far *infinite* cases into one of two types: those that are now *empty*, and those which remain *infinite*.
+
+A smaller issue proposed to be fixed by this RFC is the naming of the Range bounds `min` and `max`. Especially when an implementation allows for descending Ranges like Pony's does -- e.g. `Range(-10, -20, -1)`, naming the first parameter `min` and the second `max` is clearly mathematically wrong. Therefore if accepted, in the actual implementation these will be renamed to `start` and `end`. However, to not confuse things, they will still be named `min, max` throughout here.
```

Comment:
Renaming parameters could be a breaking change (when named arguments are used) but in the present case, current parameter names are misleading and this renaming is expected.

Reply:
Agreed.